flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/FlinkKafkaInternalProducer.java

```diff
@@ -72,12 +74,17 @@ private static Properties withTransactionalId(
         return props;
     }
 
+    public long getPendingRecordsCount() {
+        return pendingRecords.get();
+    }
+
     @Override
     public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
         if (inTransaction) {
             hasRecordsInTransaction = true;
         }
-        return super.send(record, callback);
+        pendingRecords.incrementAndGet();
+        return super.send(record, new TrackingCallback(callback));
```

Comment:
This change creates a new callback with every `send`. Since the callback being passed in our codebase is mostly constant, we should add a simple cache like `new LRUMap(3);`. The number is kind of arbitrary and 1 should work already. The most important part is that it shouldn't grow boundless or we get the next memory leak if I overlooked a dynamic usage ;).

Reply:
just want to check, are we proposing putting instance of TrackingCallback into the rotating cache? wouldn't that caused the previous callback that might not have been invoked to be ignored?

Reply:
I can't quite follow. I was proposing to use

`return super.send(record, callbackCache.computeIfAbsent(callback, TrackingCallback::new));`

So we have 3 cases:
- New callback, wrap in `TackingCallback` and cache.
- Existing callback (common case), retrieve existing callback and use it.
- Remove existing `TackingCallback` from cache if full.

In all cases, both the TackingCallback and the original callback will be invoked. The only difference to the code without cache is that we avoiding creating extra TrackingCallback instances around the same original callback.